File: android/src/main/java/xiuyue/tech/plugins/clipboard/YueClipboard.java
package xiuyue.tech.plugins.clipboard;

import android.content.ClipData;
import android.content.ClipboardManager;
import android.content.Context;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.net.Uri;
import android.util.Base64;
import android.util.Log;
import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.net.HttpURLConnection;
import java.net.URL;
import java.util.ArrayList;
import java.util.List;

public class YueClipboard {

    private static final String TAG = "YueClipboard";
    private ClipboardManager clipboardManager;
    private Context context;

    public YueClipboard(Context context) {
        this.context = context;
        this.clipboardManager = (ClipboardManager) context.getSystemService(Context.CLIPBOARD_SERVICE);
    }

    public void write(String text, String imageData, String url) throws ClipboardException {
        try {
            if (text != null && !text.isEmpty()) {
                ClipData clip = ClipData.newPlainText("text", text);
                clipboardManager.setPrimaryClip(clip);
                Log.d(TAG, "Text written to clipboard");
            } else if (url != null && !url.isEmpty()) {
                ClipData clip = ClipData.newPlainText("url", url);
                clipboardManager.setPrimaryClip(clip);
                Log.d(TAG, "URL written to clipboard");
            } else if (imageData != null && !imageData.isEmpty()) {
                // For images, we'll store as URI or convert to text representation
                // Android clipboard primarily handles text, URI, and Intent data
                // For images, we can store the data URL as text
                ClipData clip = ClipData.newPlainText("image", imageData);
                clipboardManager.setPrimaryClip(clip);
                Log.d(TAG, "Image data written to clipboard as text");
            } else {
                throw new ClipboardException("No valid content provided to write to clipboard");
            }
        } catch (Exception e) {
            throw new ClipboardException("Failed to write to clipboard: " + e.getMessage(), e);
        }
    }

    public ClipboardReadResult read() throws ClipboardException {
        try {
            if (!clipboardManager.hasPrimaryClip()) {
                return new ClipboardReadResult("", "text");
            }

            ClipData clip = clipboardManager.getPrimaryClip();
            if (clip == null || clip.getItemCount() == 0) {
                return new ClipboardReadResult("", "text");
            }

            ClipData.Item item = clip.getItemAt(0);
            String text = "";
            String type = "text";

            // Check for text content
            if (item.getText() != null) {
                text = item.getText().toString();
                
                // Determine content type based on content
                if (text.startsWith("data:image/")) {
                    type = "image";
                } else if (text.matches("^https?://.*")) {
                    type = "url";
                } else {
                    type = "text";
                }
            } else if (item.getUri() != null) {
                // Handle URI content
                Uri uri = item.getUri();
                text = uri.toString();
                type = "url";
            }

            Log.d(TAG, "Read from clipboard: " + type);
            return new ClipboardReadResult(text, type);
        } catch (Exception e) {
            throw new ClipboardException("Failed to read from clipboard: " + e.getMessage(), e);
        }
    }

    public boolean hasContent() {
        try {
            return clipboardManager.hasPrimaryClip() && 
                   clipboardManager.getPrimaryClip() != null &&
                   clipboardManager.getPrimaryClip().getItemCount() > 0;
        } catch (Exception e) {
            Log.e(TAG, "Failed to check clipboard content: " + e.getMessage());
            return false;
        }
    }

    public void clear() throws ClipboardException {
        try {
            ClipData clip = ClipData.newPlainText("", "");
            clipboardManager.setPrimaryClip(clip);
            Log.d(TAG, "Clipboard cleared");
        } catch (Exception e) {
            throw new ClipboardException("Failed to clear clipboard: " + e.getMessage(), e);
        }
    }

    public List<String> getAvailableTypes() {
        List<String> types = new ArrayList<>();
        
        try {
            if (!clipboardManager.hasPrimaryClip()) {
                return types;
            }

            ClipData clip = clipboardManager.getPrimaryClip();
            if (clip == null || clip.getItemCount() == 0) {
                return types;
            }

            ClipData.Item item = clip.getItemAt(0);

            if (item.getText() != null) {
                String text = item.getText().toString();
                if (text.startsWith("data:image/")) {
                    types.add("image");
                } else if (text.matches("^https?://.*")) {
                    types.add("url");
                } else {
                    types.add("text");
                }
            }

            if (item.getUri() != null) {
                if (!types.contains("url")) {
                    types.add("url");
                }
            }

            Log.d(TAG, "Available clipboard types: " + types.toString());
        } catch (Exception e) {
            Log.e(TAG, "Failed to get available types: " + e.getMessage());
        }

        return types;
    }

    // Helper class for read results
    public static class ClipboardReadResult {
        public final String value;
        public final String type;

        public ClipboardReadResult(String value, String type) {
            this.value = value;
            this.type = type;
        }
    }

    // Custom exception for clipboard operations
    public static class ClipboardException extends Exception {
        public ClipboardException(String message) {
            super(message);
        }
        public ClipboardException(String message, Throwable cause) {
            super(message, cause);
        }
    }

    // Original echo method (kept for reference)
    public String echo(String value) {
        Log.i("Echo", value);
        return value;
    }
}

File: ios/Sources/YueClipboardPlugin/YueClipboard.swift
import Foundation
import UIKit

@objc public class YueClipboard: NSObject {

    // MARK: - Clipboard Operations

    @objc public func write(text: String?, image: String?, url: String?) throws {
        let pasteboard = UIPasteboard.general
        
        // Clear the pasteboard first
        pasteboard.items = []
        
        if let text = text, !text.isEmpty {
            // Write text to clipboard
            pasteboard.string = text
            print("Text written to clipboard")
        } else if let url = url, !url.isEmpty {
            // Write URL to clipboard
            pasteboard.string = url
            print("URL written to clipboard")
        } else if let imageData = image, !imageData.isEmpty {
            // Handle image data (base64 or URL)
            if imageData.hasPrefix("data:image/") {
                // Handle base64 data URL
                if let data = dataFromDataURL(imageData) {
                    if let uiImage = UIImage(data: data) {
                        pasteboard.image = uiImage
                        print("Image written to clipboard from data URL")
                    } else {
                        // Fallback to storing as text
                        pasteboard.string = imageData
                        print("Image data written to clipboard as text (invalid image data)")
                    }
                } else {
                    throw ClipboardError.invalidImageData
                }
            } else if let imageURL = URL(string: imageData) {
                // Handle image URL - download and convert to image
                do {
                    let data = try Data(contentsOf: imageURL)
                    if let uiImage = UIImage(data: data) {
                        pasteboard.image = uiImage
                        print("Image written to clipboard from URL")
                    } else {
                        throw ClipboardError.invalidImageData
                    }
                } catch {
                    throw ClipboardError.failedToLoadImage(error.localizedDescription)
                }
            } else {
                // Store as text if not a valid URL or data URL
                pasteboard.string = imageData
                print("Image data written to clipboard as text")
            }
        } else {
            throw ClipboardError.noValidContent
        }
    }

    @objc public func read() throws -> ClipboardReadResult {
        let pasteboard = UIPasteboard.general
        
        // Check for image first
        if let image = pasteboard.image {
            // Convert image to base64 data URL
            if let data = image.pngData() {
                let base64String = data.base64EncodedString()
                let dataURL = "data:image/png;base64,\(base64String)"
                print("Read image from clipboard")
                return ClipboardReadResult(value: dataURL, type: "image")
            }
        }
        
        // Check for text/URL
        if let string = pasteboard.string {
            let urlPattern = "^https?://.*"
            let regex = try NSRegularExpression(pattern: urlPattern, options: .caseInsensitive)
            let range = NSRange(location: 0, length: string.utf16.count)
            
            if regex.firstMatch(in: string, options: [], range: range) != nil {
                print("Read URL from clipboard")
                return ClipboardReadResult(value: string, type: "url")
            } else {
                print("Read text from clipboard")
                return ClipboardReadResult(value: string, type: "text")
            }
        }
        
        // No content found
        return ClipboardReadResult(value: "", type: "text")
    }

    @objc public func hasContent() -> Bool {
        let pasteboard = UIPasteboard.general
        return pasteboard.hasStrings || pasteboard.hasImages
    }

    @objc public func clear() throws {
        let pasteboard = UIPasteboard.general
        pasteboard.items = []
        print("Clipboard cleared")
    }

    @objc public func getAvailableTypes() -> [String] {
        let pasteboard = UIPasteboard.general
        var types: [String] = []
        
        if pasteboard.hasImages {
            types.append("image")
        }
        
        if pasteboard.hasStrings, let string = pasteboard.string {
            let urlPattern = "^https?://.*"
            do {
                let regex = try NSRegularExpression(pattern: urlPattern, options: .caseInsensitive)
                let range = NSRange(location: 0, length: string.utf16.count)
                
                if regex.firstMatch(in: string, options: [], range: range) != nil {
                    types.append("url")
                } else {
                    types.append("text")
                }
            } catch {
                types.append("text") // Default to text if regex fails
            }
        }
        
        print("Available clipboard types: \(types)")
        return types
    }

    // MARK: - Helper Methods

    private func dataFromDataURL(_ dataURL: String) -> Data? {
        let components = dataURL.components(separatedBy: ",")
        if components.count == 2 {
            let base64String = components[1]
            return Data(base64Encoded: base64String)
        }
        return nil
    }

    // MARK: - Error Handling

    enum ClipboardError: Error, LocalizedError {
        case noValidContent
        case invalidImageData
        case failedToLoadImage(String)

        public var errorDescription: String? {
            switch self {
            case .noValidContent:
                return "No valid content provided to write to clipboard."
            case .invalidImageData:
                return "Invalid image data provided."
            case .failedToLoadImage(let message):
                return "Failed to load image: \(message)"
            }
        }
    }

    // MARK: - Result Helper

    @objc public class ClipboardReadResult: NSObject {
        @objc public let value: String
        @objc public let type: String
        
        @objc public init(value: String, type: String) {
            self.value = value
            self.type = type
        }
    }

    // Original echo method (kept for reference)
    @objc public func echo(_ value: String) -> String {
        print(value)
        return value
    }
}
